new feature: userid for PGString

image357 opened the pull request:
I've managed to get user defined ids working without breaking the possibility of linking entities via "iot.id".
The key was to alter "entityExistsOrCreate" inside EntityInserter.
It now checks, if an entity could be a link entity before it tries to create it.

There is a new option in context.xml called "persistence.defaultIdMode" which can take on three values:
* 1: enforce standard default id mode (no user defined ids allowed)
* 2: allow both, user defined ids and default ids
* 3: enfoce user defined id mode (no default ids allowed)

Currently defaults to 2 if the option is undefined.
Loglevel for userids is set accordingly to that option.

There is a new class UserDefinedId, which encapsulates consistency checks / id parsing etc. for userids.
We plan to extend this class (again with options for default behavior) for extended user id handling like prefixing with entity type etc.

For instance, the follwing json is now a valid Thing definition:
```json
{
    "@iot.id": "MyFROSTUserId",
    "name": "Some Thing",
    "description": "Test JSON for a Thing",

    "properties": {
        "Deployment Condition": "Deployed in a third floor balcony",
        "Case Used": "Radiation shield"
    }
}
```

hylkevds commented:
Thanks for the PR!

There are some important issues:

1. You can not change existing change sets in the Liquibase changelog files, since this breaks existing installations. If you need to make database changes, you have to do it in a new changeset.
2. Don't change default behaviour, such as the default DB connection string, or the default persistence manager. A user updating his server already has a config file. This config file must still work, and show the same behaviour as before.
3. Instances of the Id classes created outside of the specific persistence manager package should always be created by the IdManager that is returned by the persistence manager. Code in the core must be Id-implementation independent. Currently your code breaks if I use a different persistence manager.

Then there are some code quality points:

1. Don't use magic integers, use Enums instead.
2. Use clear names:
    - the UserDefinedId class is not an ID. From the name it is not clear what this class does.
    - forcedefaultid / allowdefaultid? What do they do?
    - The setting "defaultIdMode" is not a default. That setting should probably be named idMode (idGenerationMode), with a default of "ServerGeneratedOnly".
3. Classes and public methods should have javadoc descriptions.
4. The UserDefinedId class does not need to create new ids (line 100), it already got one from the entity in the constructor (line 49). The parsing already happened long ago.
5. Bad form: `if (allowdefaultid == false) {` use `if (!allowdefaultid) {`
6. The switch statement to configure the UserDefinedId class should be in the UserDefinedId class itself.

image357 commented:
Thank you for your feedback! I will work on fixing these issues and let you know when I'm done.

hylkevds commented:
Why exactly are you changing all text columns over to varchar? It's better not to change things that are not related to the feature being implemented.

image357 commented:
I reverted the changelog back to defaults and changed only the ID types to varchar from varchar(n), i.e. removed the length constraints of those strings in a seperate changeSet. This is required because an ID will usually contain more than 36 characters.

Initially I thought that varchar would be better performance wise because it will be stored inline while text will not, e.g. I/O load will be less. However varchar has still an upper limit of 65535 characters. Hence, I didn't touch the text types this time.

image357 commented:
> You still have the Datastream.UNIT columns in your change.
Yes, these were varchar(255) types I think. Do you want to keep them on fixed length?

> PostgreSQL is not MySQL :)
Oh, I thought that these difference would be SQL universal. But you are right.

hylkevds commented:
If a change has nothing to do with the feature of the pull request, then don't include it. This is especially true for database changes, since those can break existing installations.

In the end you should probably flatten the commits into a single one, since changing things back and forth makes the commits harder to follow. As long as the changes are not merged to master you can still do that.

image357 commented:
Ok, I'll remove it and flatten the commits.

image357 commented:
I think I'm finished with your remarks.
Towards your important issues:
1. > You can not change existing change sets in the Liquibase changelog files, since this breaks existing installations. If you need to make database changes, you have to do it in a new changeset.
   
   I've added a new changeSet and tested migration from "old" databases to the new one.
2. > Don't change default behaviour, such as the default DB connection string, or the default persistence manager. A user updating his server already has a config file. This config file must still work, and show the same behaviour as before.  
   
   DB connection string and persistence manager are unchanged now.
3. > Instances of the Id classes created outside of the specific persistence manager package should always be created by the IdManager that is returned by the persistence manager. Code in the core must be Id-implementation independent. Currently your code breaks if I use a different persistence manager.  
   
   Moved that class into PGString since it was PGString-specific anyway


Towards your code quality remarks:
1. > Don't use magic integers, use Enums instead.   
   
   Done, settings use enums and strings now.
2. > Use clear names: the UserDefinedId class is not an ID. From the name it is not clear what this class does. forcedefaultid / allowdefaultid? What do they do? The setting "defaultIdMode" is not a default. That setting should probably be named idMode (idGenerationMode), with a default of "ServerGeneratedOnly".  
   
   Done, force/allowdefault id are gone. Instead settings via enums are used to decide that cases. "ServerGeneratedOnly" is default now.
3. > Classes and public methods should have javadoc descriptions.  
   
   Done
4. > The UserDefinedId class does not need to create new ids (line 100), it already got one from the entity in the constructor (line 49). The parsing already happened long ago.  
   
   IdString / Id class doesn't allow for setting new id values. Hence after parsing, a new one has to be created. However, this is currently not in use (commented out).
5. > Bad form: if (allowdefaultid == false) { use if (!allowdefaultid) {  
   
   Yes, removed it.
6. > The switch statement to configure the UserDefinedId class should be in the UserDefinedId class itself.
   
   Done


I have 4 merge commits in the branch history. Tried to rebase the whole history into a single commit but I have to manually merge all the merge commits with upstream/master. If it's ok and you are fine with the current status, I will just create a new branch and apply current HEAD to that branch, close this PR and open a clean, new PR.

image357 commented:
I've found a way to nicely squash the commits: https://stackoverflow.com/a/17141512/8088550

I will do that when there are no further remarks.

hylkevds commented:
Thanks, looks good.

My point for 4 was that you don't need to parse and create a new id object at all. You already have a parsed and ready to use Id object. The method parseId() only needs to validate the already existing id object, not create a new id object.

Since there is currently nothing actually string-id-specific in the IdGenerationHandler I'll probably refactor it into the generic package, and make subclasses that implement the parseId (or better: validateId) method.

One thing I'm not certain about is throwing an exception when an Id is supplied and the mode is ServerGeneratedOnly. Currently the user-supplied id is simply ignored, so it might be better to keep that behaviour and have checkId return false in that case.

It looks like I can also do the squash when I merge the changes, though that probably puts my name on the commit, instead of yours.

image357 commented:
> One thing I'm not certain about is throwing an exception when an Id is supplied and the mode is ServerGeneratedOnly. Currently the user-supplied id is simply ignored, so it might be better to keep that behaviour and have checkId return false in that case.

Ok, I removed the exception and added a logger warning just in case someone wants to check where @iot.id was used.

> The method parseId() only needs to validate the already existing id object, not create a new id object.

We later want to use parseId() to change the client defined id based on context, e.g. append a datestring or add a entity prefix. We also want to do that based on options given in context.xml. Validity checks should be in checkId().
That's why we have these blocks
```
if (idhandler.checkId()) {
    idhandler.parseId();
    insert.set(qd.id, idhandler.getId());
}
```
for every insertEntity() method in EntityInserter.java. The additional checkId() call inside parseId() is just for safety reason, if anyone ever tires to parse an unchecked id (e.g. null reference).

> It looks like I can also do the squash when I merge the changes, though that probably puts my name on the commit, instead of yours.

I tried the squashing of the branch history and it worked. It only contains a single commit now.

image357 commented:
Yes. I also had that split of ``useUserSuppliedId`` and ``validateUserSuppliedId`` before but currently there is nothing special to seperate them in context with the new option enum. ``useUserSuppliedId`` is nothing but a new placeholder for ``ClientGeneratedOnly`` and ``ServerAndClientGenerated`` with null reference check.

I skipped this function to have the whole switch over the enum in on place.

hylkevds commented:
The difference between `useUserSuppliedId` and `validateUserSuppliedId` is that `useUserSuppliedId` will be the same for all back-ends, while `validateUserSuppliedId` will be different for Long-ids, UUID-ids and String-ids.
This feature will have to work for all three back-ends, not just the String back-end...

image357 commented:
Ok, I'll add this then.

hylkevds commented:
Great, Thanks!
I'll look over it tomorrow, run our test-suite against it, and merge it in.
